### PR TITLE
[mqtt] HA discovery: move common configuration to base class

### PR DESCRIPTION
**Pull request by jochen314**
I know, this touches many files, but i think is if worth it.

This creates a common base class HAConfiguration for all home assistent components.
This way we do not need to repeat the common properties every time and we have a place add common functionality later (~ replacement)

We also gain the possibility to add device properties (vendor, model, firmware) to the thing, if the device specifies them.

As I do not have a working PaperUI right now, I could not fully test this step including the front-end...

### CODE REVIEW

File: addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/HAConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.thing.Thing;
+
+import com.google.gson.Gson;
+import com.google.gson.annotations.SerializedName;
+
+@NonNullByDefault
+public abstract class HAConfiguration {
+
+    private static class Config extends HAConfiguration {
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/HAConfiguration.java]: I didn't understand this static declaration on my first read through. That's a sign that there are some javaDocs missing here :)

File: addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/HAConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.thing.Thing;
+
+import com.google.gson.Gson;
+import com.google.gson.annotations.SerializedName;
+
+@NonNullByDefault
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/HAConfiguration.java]: Could you please document this class, especially the public methods. Usually we do not comment private methods / inner classes, but there is one that I like to have some docs as well.

Also: Missing author tag

**Comment by davidgraeff** [comment]
When you are ready, can you remove the WIP tag from the title?

**Comment by jochen314** [comment]
It seems, I am not able to maintain labels?

**Comment by davidgraeff** [comment]
That's alright, the label will be removed by one of the maintainers.

File: addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/ComponentFan.java [hunk 3]
```diff
@@ -28,40 +28,29 @@
...
-    static class Config {
-        protected String name = "MQTT fan";
-        protected String icon = "";
-        protected int qos = 1;
-        protected boolean retain = true;
-        protected @Nullable String state_value_template;
-        protected @Nullable String unique_id;
+    static class Config extends HAConfiguration {
+        Config() {
+            super("MQTT fan");
```
[the remark this answers is not in the document]
Comment (davidgraeff) [reply addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/ComponentFan.java]: My fault actually, but a fix is welcome

File: addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/AbstractComponent.java [hunk 4]
```diff
@@ -41,7 +41,7 @@
  * @author David Graeff - Initial contribution
  */
 @NonNullByDefault
-public abstract class AbstractComponent {
+public abstract class AbstractComponent<CONFIG extends HAConfiguration> {
```
[the remark this answers is not in the document]
Comment (jochen314) [reply addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/AbstractComponent.java]: I tend to use a different writing for template parameter to make it clearer, that it is not a normal type.
I find different conventions:
- https://docs.oracle.com/javase/tutorial/java/generics/types.html: Single capital letter
- https://google.github.io/styleguide/javaguide.html#s5.2.8-type-variable-names: Class anem followed by T

Lookin at the other openhab addons, I see mostly one letter generic parameter, with just a few expections like `org.openhab.binding.mqtt.generic.internal.tools.ChildMap<TYPE>`, `org.openhab.binding.mqtt.generic.internal.tools.DelayedBatchProcessing<TYPE>`.

I will happly change them all to one letter gernerics parameters

**Pull request merged**